### ISSUE TITLE
Support conda (python)

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#2213]
https://conda.io/docs/

**Comment by meg-hegde** [comment]
Hi, does Renovate now support conda?

**Comment by meg-hegde** [comment]
Hi, just wondering whether there are plans to add conda support soon? Alternatively, shall I try adding it using the instructions here: https://github.com/renovatebot/renovate/blob/master/docs/development/adding-a-package-manager.md?

**Comment by rarkins** [comment]
No plans, and a PR would be very welcome! I updated the doc just now to make sure it's current.

**Comment by meg-hegde** [comment]
Thank you for updating the docs - I'll give this a go when I have some time :)

**Comment by gerbenoostra** [comment]
#6969  duplicated this, thus closed it. The relevant comments from there:
**What would you like Renovate to be able to do?**

To also verify python package versions in conda environment files (`environment.yml`)

**Did you already have any implementation ideas?**
no

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

Conda environments can also include pip requirememnts, a workaround is to put those in a separate txt file, and have renovatebot check those.
`environment.yml` would then be:
```
dependencies:
- python=3.7
- jupyter
- pip
- pip:
  - -rrequirements.txt
```
This workaround however does not work for the conda packages (like the python=3.7 here, and any conda packages installed, like jupyter in this case)

**Is this a feature you'd be interested in implementing yourself?**
maybe

** Related features**
Relates to #931 , but that only implemented pip dependencies.

**Comment by rarkins** [comment]
It would be helpful if anyone can provide some public repo examples that can be tested against, as well as clarifications on file naming / file syntax. For example should we match against every `environment.yml` file in the repo?

**Comment by padmick** [comment]
From their docs https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file they reference it against environment.yml files. I know internally we use different names for our different projects (IE we name our env the same codename as the project so titan.yml ect) but if that is an issue we can look at changing them back to the conda default name.

**Comment by DavidSpek** [comment]
I'm also interested in this as the PyTorch images (and the jupyter-stack ones) use conda.

**Comment by AndreaGiardini** [comment]
We could give it a shot, starting from the datasource. I am not very familiar with typescript but I could give it a go.

The docs in the datasource require a function called ``getReleases` with input:

    `lookupName`: the package's full name including scope if present (e.g. @foo/bar)
    `registryUrls`: an array of registry Urls to try

If I understood it correctly, for something like `https://anaconda.org/conda-forge/proj/` we will have:

`lookupName` -> conda-forge/proj
`registryUrl` -> https://anaconda.org/

Am I missing something?

**Comment by DavidSpek** [comment]
I would just like to make a note that the conda-forge channel would be important to have support for (also in terms of ToS of the regular conda channel).

**Comment by rarkins** [comment]
I agree that starting with the datasource first makes best sense.

Importantly, note that Renovate doesn't yet have the concept of "platform" for datasources but it looks like that might be necessary for conda packages?

**Comment by mathbunnyru** [comment]
Jupyter docker images would also greatly benefit from this feature.
https://github.com/jupyter/docker-stacks/issues/1153

Right now, we're updating our dependencies manually, and it would be great to get rid of this maintenance burden.

**Comment by morremeyer** [comment]
Hey everyone, Anaconda engineer here. We've assembled a small group of engineers that will look into adding functionality for conda over time.

Please note that none of us is working on this full time right now, but work will be done over time.

I've started implementing a datasource for conda in #14257, any help with my testing issue and feedback in general is very welcome!

**Comment by mdehollander** [comment]
> Hey everyone, Anaconda engineer here. We've assembled a small group of engineers that is looking into adding functionality for conda over time.
> 
> Please note that none of us is working on this full time right now, but work will be done over time.
> 
> I've started implementing a datasource for conda in #14257, any help with my testing issue and feedback in general is very welcome!

Thanks for starting the work on this! I am happy to start testing, but I am not sure how since I am new to renovate. I gave it a try with adding a `renovate.json` to my repo (https://github.com/mdehollander/orochi/blob/master/renovate.json). The conda environment files are the folder `src/envs/`. But I did not manage to set it up correctly since I get this error:
```
validationMessage": "Invalid configuration option: conda, The following managers configured in enabledManagers are not supported: \"conda\"",
```
What is the best way to test and configure this? Or I am too early :smiley_cat:

**Comment by morremeyer** [comment]
@mdehollander You're not too early, but there's no **manager** implemented yet, just a **datasource**. I'll take a first shot at a manager next Friday.

You can check the datasource documentation at https://docs.renovatebot.com/modules/datasource/#conda-datasource.

If you want to use it right now, here's an example for how to do so. In your `environment.yaml`, add a comment that annotates the line for renovate:

```
name: your-project
channels:
  - defaults
dependencies:
  - pytest
  - pytest-cov
  - coverage
  # renovate datasource=conda depName=main/yapf
  - yapf==0.31.0
```

This annotates the `yapf` package for the regexManager and will query the `main` channel for the versions. You can then use the regexManager with the following configuration:

```json
{
  "reviewersFromCodeOwners": "true",
  "regexManagers": [
    {
      "description": "Upgrade conde dependencies",
      "fileMatch": [
        "(^|/)environment.yml$"
      ],
      "matchStrings": [
        "# renovate datasource=conda\\sdepName=(?<depName>.*?)\\s+- [a-z0-9]+==\"?(?<currentValue>.*)\"?"
      ],
      "datasourceTemplate": "conda"
    }
  ]
}
```

The job of the package manager is to do the discovery/annotation that I show above automatically so that you don’t need any configuration in the default case.

**Comment by mdehollander** [comment]
Thanks for the extra information and the example config. With this I managed to get a PR triggered for an update of a conda environment.
From the logs:
```
DEBUG: Matched 33 file(s) for manager regex: src/envs/amos.yaml, src/envs/antismash.yaml, src/envs/bamm.yaml, src/envs/bbmap.yaml, src/envs/bedtools.yaml, src/envs/bigscape.yaml, src/envs/bigslice.yaml, src/envs/bwa.yaml, src/envs/cat.yaml, src/envs/checkm.yaml, src/envs/concoct.yaml, src/envs/coverm.yaml, src/envs/dastool.yaml, src/envs/fraggenescan.yaml, src/envs/groopm.yaml, src/envs/khmer.yaml, src/envs/kraken.yaml, src/envs/mash.yaml, src/envs/megahit.yaml, src/envs/metabat.yaml, src/envs/minimap2.yaml, src/envs/mmgenome.yaml, src/envs/mmgenome_prepare.yaml, src/envs/pigz.yaml, src/envs/prodigal.yaml, src/envs/quast.yaml, src/envs/report.yaml, src/envs/samtools.yaml, src/envs/seqtk.yaml, src/envs/spades.yaml, src/envs/tree.yaml, src/envs/vamb.yaml, src/envs/vsearch.yaml

DEBUG: 2 flattened updates found: bioconda/spades, bioconda/vsearch
DEBUG: Returning 2 branch(es)
DEBUG: Fetching changelog: https://github.com/ablab/spades (3.14.0 -> 3.15.4)
```

And 2 PRs for 2 packages that I enabled: https://github.com/mdehollander/orochi/pulls

![image](https://user-images.githubusercontent.com/262872/156830355-867da2d7-66b9-4788-bf8a-4b65388d48dd.png)

To get the environment files in subfolders recognized I changed the regular expression in the config to:
```
      "fileMatch": [
        "^(?:src/envs/)?\\w+\\.yaml$"
      ],
```
Looks very promising! Thanks! Looking forward for a manager :)

**Comment by morremeyer** [comment]
Unfortunately, I didn’t get around to it yet as I needed to prioritize other topics. I’m estimating to get some work in for this in about 3 weeks.

**Comment by PhilipAbed** [comment]
we have the Versioning and Datasource for python, 
we just need to check for `environment.yml` file and extract dependencies and it should work with renovate normally

In fact we can do this using regex managers

**Comment by morremeyer** [comment]
@PhilipAbed I'm not sure if the python datasource supports the conda versioning spec.

_Everyone, here's what I *thought* I had posted back in May when I last got to work on this (time really flies currently):_

We worked on this over the last few days, so here is a summary and the plan for how I’m thinking about proceeding here.

There is no working manager yet that could be tested as the focus was on two things:

* Ensure that we only create *correct* updates
* Start simple, build support for more complex cases later

As the conda [MatchSpec](https://github.com/conda/conda/blob/0ce67ff01354667d6cf8c956839a401f42e41bde/conda/models/match_spec.py#L69) enables complex definitions of package versions, it’s vital to parse them correctly according to the specification so that we don’t create wrong updates.

A MatchSpec can match on practically every attribute of a package, so we need to focus on the common cases that can be correctly implemented without hacking the renovate core.

I discussed a lot of ideas with and got input by @rarkins, big thank you for that!

Here’s what is planned to be supported in the first iteration of a conda manager:

* Specifications of dependencies that include channel, subdir, name, version. Check the MatchSpec documentation linked above for details on the fields.

The first iteration will have at least the following limitations:

* No support for channel priority, it will behave like conda with channel_priority: disabled
* Dependencies with any of the following fields specified will be ignored:
* `namespace` (this is not used yet in conda, but ignoring it will prevent wrong updates)
* `build`
* any keyword argument
* The key `pip` in the dependencies will be ignored if it is a list
* No support for the nonstandard keys introduced by [conda-lock](https://github.com/conda-incubator/conda-lock).

This means that specifications like the following should be supported:

* `conda-forge::some-package=1.0`
* `numpy=1.2.3`

More complex cases can then be added over time.

_I will be working on this the week after next (likely 2022-08-10 and 2022-08-11), I'm happy about any feedback here!_

**Comment by PhilipAbed** [comment]
Thanks for the update, looks promising! good work

**Comment by srilman** [comment]
What is the status of the PR? I see that the last update was a couple of months ago. Would be happy to help get it to the finish line!

**Comment by viceice** [comment]
there's no more work from PR author after suggestions. so he's maybe no longer working on it. feel free to take it over by create a new PR with requested changes.

**Comment by srilman** [comment]
Sounds good, thanks!

**Comment by morremeyer** [comment]
Hey everyone!

Sorry for the silence over the last months. We still have this on our agenda, but didn't get to work on it in recent weeks.

Working on implementing the manager, we realized that we will first need to implement conda versioning to implement a manager that works reliably and consistently.

We definitely have not abandoned this, but as stated initially we cannot give guarantees for progress here. 

If anyone else wants to contribute to this, we're more than happy to put in our feedback for PRs.

**Comment by pavelzw** [comment]
Maybe it could make sense looking into [pixi](https://github.com/prefix-dev/pixi). They use the conda universe but in a somewhat structured way. Pixi projects always use a `pixi.toml` and have a `pixi.lock` which contain the lockfile. Somewhat similar to poetry but for conda.
Probably it's easier to implement this into dependabot since this has a clearly defined project structure.

**Comment by viceice** [comment]
@pavelzw please open a new discussion and reference this issue.

**Comment by HonkingGoose** [comment]
Hi @morremeyer :wave: 

Do you want to try (again) to get `conda` support into Renovate?

You tried getting `conda` support into Renovate with this PR from 2022:

- #17156

We closed your PR because it was getting stale. You also said that you now have a better plan for the second attempt at the PR. I don't know if you still know about the better plan, and still want to try again. :wink:

**Comment by morremeyer** [comment]
I‘m still planning to do this, but also still can not make any promises as to when. 

But this is definitely not forgotten!

**Comment by baszalmstra** [comment]
I have been looking at this for https://pixi.sh . 

@morremeyer You mentioned that you "realized that we will first need to implement conda versioning to implement a manager that works reliably and consistently". I was wondering if it would help if we package [`rattler`](https://github.com/conda-incubator/rattler) (which has a full matchspec implementation) as a wasm package so we can use it here? I think packaging javascript bindings that way would be relatively straightforward.

**Comment by morremeyer** [comment]
@baszalmstra That depends on what the maintainer's opinion on packaging other things. Our past discussions on that topic have resulted in "prefer an implementation in renovate".

I don't think that would be overly complex, I just did not get to work on it yet.

**Comment by baszalmstra** [comment]
Are you planning on implementing the entire version specification with all edge cases? From my experience this can get pretty hairy. And are you planning to implement both version spec and match spec? 

@rarkins What is your stance on depending on external packages for these things?

**Comment by rarkins** [comment]
I'm fine with using any npm package if it's cross-platform

**Comment by morremeyer** [comment]
@baszalmstra I would start implementing the most common cases and then extending it.

However, if we don't have to do that, it makes the versioning a lot more simple.

Then, what would be left would be possible improvements to the datasource and the implementation of a manager. 
I did not get any further in planning any of these since I updated this issue the last time, so any proposals or requests for feedback are welcome to me!

**Comment by viceice** [comment]
I'm fine with a wasm module which is distributed as npm package, so we only need a slim JavaScript wrapper for the versioning interface implementation.

**Comment by trim21** [comment]
`rattler` or it's version crate rattler_conda_types has a unconditionally requirements on tokio so doesn't build wasm.

**Comment by baszalmstra** [comment]
I have a working WASM build and will make a PR soon that exposes at least the version and matchspec functionality needed for renovate.

**Comment by trim21** [comment]
> I have a working WASM build and will make a PR soon that exposes at least the version and matchspec functionality needed for renovate.

Thanks! really glad to know. Very much looking forward to it.

I guess you downvote me because you I think it doesn't build on wasm and you have a working WASM build...

Yes, the unconditionally requirements on tokio was added after your checkout your wasm branch from trunk. 
It doesn't build on latest version of rattler_conda_types, but build with old version is fine.

**Comment by baszalmstra** [comment]
> I guess you downvote me because you I think it doesn't build on wasm and you have a working WASM build...

I apologize, I meant to upvote instead of downvote!

Ill try to open the PR in the next few hours! But at home with a newborn so no promises. 😄

**Comment by baszalmstra** [comment]
I [created the PR](https://github.com/conda/rattler/pull/1079).

and uploaded an initial version to npmjs.org: https://www.npmjs.com/package/@baszalmstra/rattler

**Comment by trim21** [comment]
I start working on conda versioning based  on wasm npm package. 

I'll only add pixi support in the future but not conda support.

**Comment by trim21** [comment]
A single conda dep can result into multiple conda packages from different channels, how should we support it?

Should manager generate multiple packages for single deps and let data source query them all?

**Comment by trim21** [comment]
there is a limitation in current implement of conda source，, it doesn't supprt third part channels.

But conda registry by default doesn't have API supoort for single package, do we need to download more than 6 or huge (maybe >100mb) repodata.json from single channel to get info of single package, which is not very ideal.

**Comment by baszalmstra** [comment]
The files are often also served used zstd compression which makes them much smaller (20mb). There is also an accept conda enhancement proposal called "sharded repodata" that splits the repodata into individual files per package but it is at the moment only supported by channels on prefix.dev.

rattler provides the so called repodata gateway object which is an object that hides all this complexity and allows one to simply query for the repodata of a specific package. The gateway then figures out what the most efficient way of fetching the data is, as well as caching all of it. 

I have been working on adding this to the javascript bindings of rattler with some good results but its not done yet.

For the time being, I recommend you use the zstd compressed files.

**Comment by pavelzw** [comment]
xref https://github.com/prefix-dev/pixi/discussions/3222#discussioncomment-12332106

**Comment by baszalmstra** [comment]
I assume renovate uses the fetch api, which is what the rust code will also use.

**Comment by trim21** [comment]
> I assume renovate uses the fetch api, which is what the rust code will also use.

It have a wrapped http request client to be used in all data source

https://github.com/renovatebot/renovate/blob/2c6a5001ba11a105963b32c51e0843777e124627/lib/modules/datasource/conda/index.ts#L55

**Comment by baszalmstra** [comment]
I see, yeah that complicates things. We can probably make this work but it will make things more complicated. From what I understand the `reqwest` crate which is used by rattler calls back into javascript to call the `fetch` API. I assume we can also inject another method to do this manually and allow overriding the client?

But I believe the fetch API also does caching, so as long as renovate is not requesting the same URLs (which doesnt make a lot of sense when you use the gateway) I dont think its that bad of a problem.

Ill not be implementing a custom fetch API in the initial version of the rattler gateway API. Would be happy to accept PRs though!

**Comment by trim21** [comment]
> The files are often also served used zstd compression which makes them much smaller (20mb). There is also an accept conda enhancement proposal called "sharded repodata" that splits the repodata into individual files per package but it is at the moment only supported by channels on prefix.dev.
> 
> rattler provides the so called repodata gateway object which is an object that hides all this complexity and allows one to simply query for the repodata of a specific package. The gateway then figures out what the most efficient way of fetching the data is, as well as caching all of it.
> 
> I have been working on adding this to the javascript bindings of rattler with some good results but its not done yet.
> 
> For the time being, I recommend you use the zstd compressed files.

It's not only the file size but also memory usage. for example, parsing `conda-forge/linux-64/repodata.json` will take up to 1G memory


```js
import * as fs from 'node:fs';

const file = fs.readFileSync(`./conda-forge/linux-64/repodata.json`, 'utf8');

const obj = JSON.parse(file);

console.log(process.memoryUsage());

const _ = obj;
```

```js
{
  rss: 1052028928,
  heapTotal: 1012838400,
  heapUsed: 983224232,
  external: 1691001,
  arrayBuffers: 10475
}
```

**Comment by trim21** [comment]
does anaconda has sharded_repodata now?

**Comment by baszalmstra** [comment]
The memory usage is significantly reduced by using the gateway. It does not parse the entire file as JSON but only cleverly parses the parts from the repodata that it actually needs. It does however need all the bytes in memory. With sharded repodata this problem is also mitigated.

> does anaconda has sharded_repodata now?
Unfortunately not yet.

**Comment by trim21** [comment]
> I see, yeah that complicates things. We can probably make this work but it will make things more complicated. From what I understand the `reqwest` crate which is used by rattler calls back into javascript to call the `fetch` API. I assume we can also inject another method to do this manually and allow overriding the client?
> 
> But I believe the fetch API also does caching, so as long as renovate is not requesting the same URLs (which doesnt make a lot of sense when you use the gateway) I dont think its that bad of a problem.
> 
> Ill not be implementing a custom fetch API in the initial version of the rattler gateway API. Would be happy to accept PRs though!

I think renovate also support auth config for each http host, which is supported by `this.http` here.

**Comment by trim21** [comment]
I'd like to suggest we rename current conda source to ananconda, since it only work with anaconda repo but not generic conda repo

**Comment by rarkins** [comment]
> I'd like to suggest we rename current `conda` source to `ananconda`, since it only work with anaconda repo but not generic conda repo

It depends.

Assuming that non-anaconda registries will be supported in future, would they be best added to the existing datasource, or to a separate one?

If the anaconda API is close to identical to the non-anaconda Conda APIs, then it should be the same datasource (like we do with `docker` datasource).

**Comment by trim21** [comment]
> Assuming that non-anaconda registries will be supported in future, would they be best added to the existing datasource, or to a separate one?

I think a separate one.

> If the anaconda API is close to identical to the non-anaconda Conda APIs, then it should be the same datasource (like we do with `docker` datasource).

It's not very close. non-anaconda conda repo doesn't even have API except repodata.json.

This is kind like pypi simple api vs pypi json api, but the differentce is that anaconda json api is not part of conda repo sepc.

**Comment by trim21** [comment]
It's kind like git tags / github tags

**Comment by rarkins** [comment]
The next challenge is that although we could rename `datasource/conda` to `datasource/anaconda`, and add migration code so that any existing config for `conda` is now massaged to `anaconda`, this concept would then break if we added back a `conda` datasource. Then we wouldn't know if user config was referring to `conda` (new) or the older conda/anaconda.

**Comment by trim21** [comment]
> The next challenge is that although we could rename `datasource/conda` to `datasource/anaconda`, and add migration code so that any existing config for `conda` is now massaged to `anaconda`, this concept would then break if we added back a `conda` datasource. Then we wouldn't know if user config was referring to `conda` (new) or the older conda/anaconda.

oops, I forget there is regex manager, someone is already using it.

**Comment by baszalmstra** [comment]
We could call the new one simply `conda-repodata` or something along those lines? or `conda-channel`? Naming is hard..

**Comment by trim21** [comment]
conda-channel looks good abd make more sence.

**Comment by trim21** [comment]
> conda-channel looks good and make more sence.

**Comment by rarkins** [comment]
> It's kind like git tags / github tags

Funny that you mention that. I have plans for `git-tags` to contain the logic which identifies "oh this is a github repository - let's use the github-tags datasource instead". Otherwise you force each manager to have to implement that logic. Similar with `conda` - it could be a single datasource but dispatch logic separately. Docker isn't the only one like that - it's quite common for the default registry in ones like PyPI or Cargo to have specific implementations.

**Comment by trim21** [comment]
I was actually just thinking the same think.

The manager that produce package using conda data source should prepare the full registry url, for conda-forge with anaconda it should be `https://conda.anaconda.org/conda-forge/`

Then we parse registry url in conda data source to decide we use api.anaconda or graphql of prefix.dev, or we use generic conda logic for `repodata.json`

**Comment by trim21** [comment]
Another problem: there is no way to know if a version is yanked. So you will encounter this:

renovatebot will try to update a package to yanked version and you got a broken manifest file, package manager tell you it can't find available distribution

![Image](https://github.com/user-attachments/assets/1a906bba-7397-43ab-9df1-e8aa2609ad9a)

![Image](https://github.com/user-attachments/assets/c14fc03d-1a4c-4624-9868-57d7ea9a785d)

**Comment by baszalmstra** [comment]
If you use the prefix graphql API you can use the `yankedReason` to determine if a package is yanked.

If you use repodata.json, yanked entries should be under the `removed` key.

Does that help?

**Comment by trim21** [comment]
https://github.com/renovatebot/renovate/pull/34646 should work for most case of pixi， now I just need to get it merged 😄

**Comment by trim21** [comment]
Just get support for pixi manager merged.

So some notes for conda manager (environment.yml)：

Each package in `environment.yml` is a `MatchSpec`, and if you don't want to implement it in js again, you may choose to expose it from js-ratter. 

But rattler's `MatchSpec` has a normalization **at parsing step** and which currenlly it can't be disabled. So you will need to write some rust code for it.

**Comment by julian-belina** [comment]
Is there an update regarding support for environment.yml files?

**Comment by trim21** [comment]
blocked by https://github.com/conda/rattler/issues/1142  , or you can write a js implement
